Copy functions

alimanfoo opened the pull request:
This PR adds some convenience functions for copying data between zarr groups or stores, or for copying data between h5py and zarr. The following new functions are proposed:

* ``zarr.copy(source, dest, ...)`` - copy the ``source`` h5py or zarr group or array into the ``dest`` h5py or zarr group.
* ``zarr.copy_all(source, dest, ...)`` - copy all children of the ``source`` h5py or zarr group into the ``dest`` h5py or zarr group.
* ``zarr.copy_store(source, dest, ...)`` - copy all (key, value) pairs from the ``source`` store to the ``dest`` store.

Further details and examples are given in the function docstrings.

Resolves #87, resolves #113, resolves #137.

TODO:
* [x] consider using require_group to allow for destination groups already existing
* [x] test coverage up
* [x] address all @jakirkham review comments
* [x] tutorial section
* [x] release notes

POSTPONED:
* consider Group.copy() method?
* consider ``delete=True/False`` option analogous to rsync?

alimanfoo commented:
Hi @jakirkham, hopefully this addresses several requirements related to copying data, including copying between different zarr stores, and copying data between h5py and zarr. This is the last piece of work I was planning to include in v2.2 before release candidate. If you get a chance to take a look, any comments welcome.

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
         return LazyLoader(grp)
+
+
+class _LogWriter(object):
```
This is neat. I wonder if we should start using it with other things.

Also might be worth giving [this]( https://docs.python.org/3/howto/logging-cookbook.html#using-a-context-manager-for-selective-logging ) a look.

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
+                kws.setdefault('shuffle', True)
+            else:
+                # zarr -> zarr; preserve compression options by default
+                kws.setdefault('compressor', source.compressor)
```
Should we pick up `filters` as well? Also what about things like `fill_value` and `order`? Anything else like this that we might need to get?

alimanfoo replied:
Good catch, I've add tests and implementation to pass through filters, fill_value and order when copying zarr-to-zarr.

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
+                kws.setdefault('compressor', source.compressor)
+
+        # create new dataset in destination
+        ds = dest.create_dataset(name, shape=source.shape, dtype=source.dtype, **kws)
```
Thoughts on raising if `dest` is a Datastet/Array? Admittedly this would happen already, but it might be worth providing a cleaner error than `AttributeError` and a better message.

alimanfoo replied:
Now raises ValueError if dest is not a group.

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
+
+        # copy data - N.B., if dest is h5py this will load all data into memory
+        log('{} -> {}'.format(source.name, ds.name))
+        ds[:] = source
```
Might be worth trying [`read_direct`]( http://docs.h5py.org/en/latest/high/dataset.html#Dataset.read_direct ) to avoid memory issues. Not sure how well it works with a non-NumPy array, but it is an easy thing to explore that should help this issue.

alimanfoo replied:
The logic here has been rewritten to copy chunk-by-chunk.

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
+    elif root or not shallow:
+        # copy a group
+
+        # creat new group in destination
```
nit: `creat` -> `create`

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
+        # copy a group
+
+        # creat new group in destination
+        grp = dest.create_group(name)
```
Thoughts on using `require_group` instead?

alimanfoo replied:
Done.

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
+                  without_attrs=without_attrs, **create_kws)
+
+
+def tree(grp, expand=False, level=None):
```
👍

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
+                # zarr -> h5py; use some vaguely sensible defaults
+                kws.setdefault('compression', 'gzip')
+                kws.setdefault('compression_opts', 1)
+                kws.setdefault('shuffle', True)
```
Yeah this seems like the only reasonable choice (besides no compression) to ensure compatibility in a wide array of situations. After all users can choose something different if they would like.

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
+    grp : Group
+        Zarr or h5py group.
+    expand : bool, optional
+        Only relevant for HTML representation. If True, tree will be fully expanded.
```
Sorry this is a bit off topic and I may have forgotten the answer, do we warn if `expand` is set to `True` for the text representation? If not, we might want to look into it.

jakirkham replied:
Revisiting this, it doesn't look like this would be easy to do as we don't know what representation will be used when `expand` is set.

jakirkham commented on zarr/convenience.py:
```diff
@@ -348,3 +350,392 @@ def load(store):
+        # copy a group
+
+        # creat new group in destination
+        grp = dest.create_group(name)
```
Before we starting creating groups or datasets/Arrays, we may want to do an initially pass of `dest` to make sure there will be no conflicts when writing. That way we avoid doing a partial copy and then erroring out with incomplete state.

jakirkham commented on zarr/tests/test_convenience.py:
```diff
@@ -85,3 +88,235 @@ def test_lazy_loader():
+    assert len(dest) == 2
+    assert 'foo' in dest
+    assert 'bar/baz' not in dest
+    assert 'bar/qux' in dest
```
Would be good to have some error tests? For example, what happens when a Group or Dataset/Array is already present? Do we have any incomplete copies for cases like this or is dest left unchanged?

jakirkham commented on zarr/tests/test_convenience.py:
```diff
@@ -85,3 +88,235 @@ def test_lazy_loader():
+        assert a.compressor == spam.compressor
+    assert_array_equal(a[:], spam[:])
+    assert 'foo' not in eggs
+    assert 'bar' not in eggs
```
Would be good to have some error tests here as well. For example, what happens when a Group or Dataset/Array is already present? Do we have any incomplete copies for cases like this or is dest left unchanged?

alimanfoo commented:
Thanks @jakirkham, all great comments.

On Sat, 9 Dec 2017 at 21:47, jakirkham <notifications@github.com> wrote:

> *@jakirkham* commented on this pull request.
> ------------------------------
>
> In zarr/tests/test_convenience.py
> <https://github.com/alimanfoo/zarr/pull/217#discussion_r155933070>:
>
> > +
> +    # copy group, shallow
> +    dest = new_dest()
> +    copy(source, dest, name='eggs', shallow=True)
> +    assert 'eggs' in dest
> +    eggs = dest['eggs']
> +    assert 'spam' in eggs
> +    a = eggs['spam']
> +    assert a.dtype == spam.dtype
> +    assert a.shape == spam.shape
> +    assert a.chunks == spam.chunks
> +    if hasattr(a, 'compressor') and hasattr(spam, 'compressor'):
> +        assert a.compressor == spam.compressor
> +    assert_array_equal(a[:], spam[:])
> +    assert 'foo' not in eggs
> +    assert 'bar' not in eggs
>
> Would be good to have some error tests here as well. For example, what
> happens when a Group or Dataset/Array is already present? Do we have any
> incomplete copies for cases like this or is dest left unchanged?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/alimanfoo/zarr/pull/217#pullrequestreview-82336158>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAq8QtGmb7FpoubiNoHDr_4orlLRtyyOks5s-v_tgaJpZM4Q7zZj>
> .
>
-- 
Alistair Miles
Head of Epidemiological Informatics
Centre for Genomics and Global Health <http://cggh.org>
Big Data Institute Building
Old Road Campus
Roosevelt Drive
Oxford
OX3 7LF
United Kingdom
Phone: +44 (0)1865 743596
Email: alimanfoo@googlemail.com
Web: http://a <http://purl.org/net/aliman>limanfoo.github.io/
Twitter: https://twitter.com/alimanfoo

jakirkham commented:
Thanks for tackling this @alimanfoo. Very excited to see this coming together. Also excited to see that 2.2 is fast approaching.

Tried to add some comments and questions above. The biggest things on my mind ATM can be summarized in three questions. First how do we ensure that copies of Datasets/Arrays are done in a memory efficient manner (as they whole thing may not fit)? Second how do we protect against copying erroring out part way and leaving the destination in an incomplete state (related how do we test this)? Third how do we handle more complex compression/filtering configurations? The last one may be a bit more of conscious choice between what is built in and what should be up to end users and could benefit from some docs/advice along these lines.

alimanfoo commented:
I have a proposal for how to handle a couple of these points.

Re copying without using too much memory, I've modified the array copy code to perform a chunk-by-chunk copy.

Re protecting against an incomplete copy, I have added two parameters ``if_exists`` and ``dry_run`` which can be used in several ways. 

If you want to check whether a copy will proceed to completion without errors, you can use ``if_exists='raise'`` (default) with ``dry_run=True``. This means an error is raised if an array already exists in the destination, and no actual copying will be done (dry run). E.g.:

```python
In [21]: source = zarr.group()

In [22]: dest = zarr.group()

In [23]: source.create_dataset('foo/bar/baz', data=np.arange(100))
Out[23]: <zarr.core.Array '/foo/bar/baz' (100,) int64>

In [24]: source.create_dataset('foo/spam', data=np.arange(1000))
Out[24]: <zarr.core.Array '/foo/spam' (1000,) int64>

In [25]: dest.create_dataset('foo/spam', data=np.arange(1000))
Out[25]: <zarr.core.Array '/foo/spam' (1000,) int64>

In [26]: zarr.copy(source['foo'], dest, log=sys.stdout, dry_run=True)
copy /foo
copy /foo/bar
---------------------------------------------------------------------------
...
ValueError: an object 'spam' already exists in destination '/foo'
```

Alternatively, if you want to replace any arrays in the destination, use ``if_exists='replace'``, e.g.:

```python
In [28]: zarr.copy(source['foo'], dest, log=sys.stdout, if_exists='replace', dry_run=True)
copy /foo
copy /foo/bar
copy /foo/spam (1000,) int64
dry run: 3 copy, 0 skip
```

Another alternative is ``if_exists='skip'``, which will skip any arrays that exist in the destination:

```python
In [29]: zarr.copy(source['foo'], dest, log=sys.stdout, if_exists='skip', dry_run=True)
copy /foo
copy /foo/bar
skip /foo/spam (1000,) int64
dry run: 2 copy, 1 skip
```

A final alternative is ``if_exists='skip_initialized'``, which will skip any arrays that exist in the destination **and** which have all chunks initialized. This provides a way to restart a copy that broke in the middle for some reason (e.g., disk full, network error, ...) without having to re-copy any array that was previously fully copied. This mode is only available when copying to zarr, as I can't find any way in h5py to query the number of chunks initialized in an array.

I still have to add more tests covering all these options, but thought I'd post this now to check this all sounds reasonable.

jakirkham commented on zarr/tests/test_convenience.py:
```diff
@@ -415,15 +571,33 @@ def temp_h5f():
+        self.source_h5py = True
+        self.dest_h5py = True
+        self.new_source = temp_h5f
+        self.new_dest = temp_h5f
```
Sorry if I overlooked it, but was there a ZarrToZarr case?

alimanfoo replied:
Yes, sorry, it's the super-class, i.e., TestCopy implements zarr-to-zarr. I could make this more obvious.

jakirkham replied:
Thanks.

jakirkham commented on zarr/tests/test_convenience.py:
```diff
@@ -85,3 +90,516 @@ def test_lazy_loader():
+
+def temp_h5f():
+    fn = tempfile.mktemp()
+    atexit.register(os.remove, fn)
```
Kind of a silly question, but should we be closing this first?

alimanfoo replied:
Do you mean closing the HDF5 file at exit, before removing the file?

jakirkham replied:
Yep.

alimanfoo replied:
Yes probably should. Easiest thing might be to call close() inside the test class tearDown rather than try to handle at exit.

jakirkham replied:
SGTM

jakirkham commented:
More generally I think your copying proposal is reasonable. Kind of would like to see default behavior perform the dry run first and then copy (if there were no exceptions in the dry run). Not sure if this would be a new option in this context or what. Though I guess this is easy for an end user to do as well. What do you think?

alimanfoo commented:
I started implementing a dry run as part of the copy, but then it seemed less complicated and more flexible if the user can have control over being able to do a dry run with whatever other options they like. Also I like the way rsync works and so have been modelling the options around that.

jakirkham commented:
Believe that you are right especially given that users are likely dealing with slow storage mediums (e.g. NFS).

alimanfoo commented:
OK, I've added a tutorial section and release notes, I think this is good to go. Happy holidays @jakirkham, I'll merge this in the new year and go for 2.2 release.

jakirkham commented:
Thanks @alimanfoo. Happy Holidays! Should be able to take another look on the 2nd if you’d like. Though am happy to trust your judgment here.

jakirkham commented on docs/tutorial.rst:
```diff
@@ -768,6 +768,110 @@ Here is an example using S3Map to read an array created previously::
+
+If you have some data in an HDF5 file and would like to copy some or all of it
+into a Zarr group, or vice-versa, the :func:`zarr.convenience.copy` and
+:func:`zarr.convenience.copyall` functions can be used. Here's an example
```
Should be `copy_all`?

alimanfoo replied:
Thanks, fixed.

jakirkham commented on docs/tutorial.rst:
```diff
@@ -768,6 +768,110 @@ Here is an example using S3Map to read an array created previously::
+    >>> source.close()
+
+If rather than copying a single group or dataset you would like to copy all
+groups and datasets, use :func:`zarr.convenience.copyall`, e.g.::
```
Should be `copy_all`?

alimanfoo replied:
Thanks, fixed.

alimanfoo commented:
Merging soon if no further comments.

jakirkham commented:
Thanks @alimanfoo. Meant to give this a closer look today, but I was too slow. 😄 That said, this looked pretty good the last time I went through it. Happy to see it in.
